### ISSUE TITLE
Rule Proposal: no-unscoped-case

### ISSUE THREAD

**Issue by abritinthebay** [eslint/eslint#7311]
**Please describe what the rule should do:**

This rule enforces using curly braces in switch case clauses to scope them. It pairs well with the no-case-declarations rule as that rule will not warn when case clauses are wrapped in lexical blocks.

This rule will specifically enforce a style that allows use of `let`/`const` statements inside case statements — even with the no-case-declarations rule enabled — which allows easier, iterative, refactoring and potentially less mutation of data.

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

``` js
// this will warn
switch (foo) {
    case 1:
        // code here
        break;
}

// this will not
switch (foo) {
    case 1: {
       // code here
        break;
    }
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It's highly complementary to another - existing - rule as well as being a fundamental code style rule for a highly used basic language construct. Given the other rule I was actually surprised that it didn't exist already.

**Comment by platinumazure** [comment]
I could get behind this.

Possibly `prefer-scoped-case` to avoid the double negative? But I'm okay with either.

**Comment by abritinthebay** [comment]
Sure, could be. ESLint has a lot of `no-something-something` rules so I was more trying to copy existing style. 

My only concern with "prefer" is that it sounds more optional though - which if you're throwing an error on this rule would be a bit wonky sounding.

No strong opinion tho ;)

**Comment by mysticatea** [comment]
Thank you for this issue.

We have received a related issue: #7193
So I think `switch-case-brace-style`  (similar to [brace-style](http://eslint.org/docs/rules/brace-style) rule) or `switch-case-curly` (similar to [curly](http://eslint.org/docs/rules/curly) rule) in core sounds nice.

**Comment by abritinthebay** [comment]
Ah and have a string option for specifying style (with a default)? Yeah, that's nice and flexible. Best of all worlds, I like that a lot.

**Comment by mysticatea** [comment]
This is concrete proposal:

# switch-case-curly

This rule enforces whether you should write braces on `case:`/`default:` clauses.
This rule has a string option.

- `"always"` (default) requires braces.
- `"never"` disallows braces.

If `"never"` is given, you can specify 2nd object option (e.g. `["never", {"allowDeclarations": true}]`):

- `allowDeclarations` allows braces if the `case:`/`default:` clause has declarations to avoid confliction with [no-case-declarations](http://eslint.org/docs/rules/no-case-declarations).

<details><summary>Examples:</summary>

## always

```js
/*eslint switch-case-curly: error */

// ✔ GOOD
switch (a) {
    case 1: {
        doSomething();
        break;
    }
    case 2:
    {
        doSomething();
        break;
    }
    default: {
        doSomething();
        break;
    }
}

// ✘ BAD
switch (a) {
    case 1:
        doSomething();
        break;

    default:
        doSomething();
        break;
}
```

## never

```js
/*eslint switch-case-curly: [error, never] */

// ✔ GOOD
switch (a) {
    case 1:
        doSomething();
        break;

    default:
        doSomething();
        break;
}

// ✘ BAD
switch (a) {
    case 1: {
        doSomething();
        break;
    }
    default: {
        doSomething();
        break;
    }
}
```

## never and allowDeclarations

```js
/*eslint switch-case-curly: [error, never, {allowDeclarations: true}] */

// ✔ GOOD
switch (a) {
    case 1:
        doSomething();
        break;

    case 2:
        const b = 1;
        doSomething(b);
        break;

    case 3: {
        const c = 1;
        doSomething(c);
        break;
    }

    default:
        doSomething();
        break;
}

// ✘ BAD
switch (a) {
    case 1: {
        doSomething();
        break;
    }
    default: {
        doSomething();
        break;
    }
}
```

</details>

----

Hmm, this rule overlaps a bit to [no-case-declarations](http://eslint.org/docs/rules/no-case-declarations).
[no-case-declarations](http://eslint.org/docs/rules/no-case-declarations) also is the rule which enforces braces on `case:`/`default:` clauses.

**Comment by abritinthebay** [comment]
It's not so much an overlap to `no-case-declarations` as complementary to it - as that rule cares specifically about vars and it specifically doesn't warn when you use curly braces.

It would likely be listed on that page as a related rule: just like `no-fallthrough` is. In fact that was the direct intent.

**Comment by kaicataldo** [comment]
Closing this as discussion has stalled and it doesn't appear there is enough interest, either from the team or from the community, to continue the discussion at this time. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. Please feel free to open a new issue with a proposal to continue this discussion if this is something you would like to continue advocating for.

**Comment by alxndr** [comment]
I wonder if this would be possible to implement using [the AST Selectors introduced in v3.18.0](http://eslint.org/blog/2017/03/eslint-v3.18.0-released).
